mgr/cephadm: adding new command to get cephadm default images

rkachach opened the pull request:
https://tracker.ceph.com/issues/72021

Adding support both in the cephadm binary as in the cephadm CLI to get default images:

```
./cephadm get-default-images | jq                                                                 ✔ 
{
  "PROMETHEUS": {
    "key": "container_image_prometheus",
    "image_ref": "quay.io/prometheus/prometheus:v2.51.0",
    "desc": "Prometheus container image"
  },
  "LOKI": {
    "key": "container_image_loki",
    "image_ref": "docker.io/grafana/loki:3.0.0",
    "desc": "Loki container image"
  },
  "PROMTAIL": {
    "key": "container_image_promtail",
    "image_ref": "docker.io/grafana/promtail:3.0.0",
    "desc": "Promtail container image"
  },

```

From cephadm CLI:

```
[ceph: root@ceph-node-0 /]# ceph orch get-default-images | jq
{
  "PROMETHEUS": {
    "key": "container_image_prometheus",
    "image_ref": "quay.io/prometheus/prometheus:v2.51.0",
    "desc": "Prometheus container image"
  },
  "LOKI": {
    "key": "container_image_loki",
    "image_ref": "docker.io/grafana/loki:3.0.0",
    "desc": "Loki container image"
  },
  "PROMTAIL": {
    "key": "container_image_promtail",
    "image_ref": "docker.io/grafana/promtail:3.0.0",
    "desc": "Promtail container image"
  },
  "NODE_EXPORTER": {
    "key": "container_image_node_exporter",
    "image_ref": "quay.io/prometheus/node-exporter:v1.7.0",
    "desc": "Node exporter container image"
  },

```



<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

phlogistonjohn reviewed (commented):
looks ok to me on first glance, but why not set an indent value on the json dump so that you don't need jq to get more human-readable output?

rkachach commented:
> looks ok to me on first glance, but why not set an indent value on the json dump so that you don't need jq to get more human-readable output?

@phlogistonjohn yeah, we can do that for sure. I'll update the code :+1:

adk3798 commented:
@rkachach we have this command already in the binary
```
[root@vm-00 ~]# cephadm list-images
[mgr]
mgr/cephadm/container_image_prometheus = registry.redhat.io/openshift4/ose-prometheus:v4.15
mgr/cephadm/container_image_loki = registry.redhat.io/openshift-logging/logging-loki-rhel8:v3.1.0
mgr/cephadm/container_image_promtail = registry.redhat.io/rhceph/rhceph-promtail-rhel9:v3.0.0
mgr/cephadm/container_image_node_exporter = registry.redhat.io/openshift4/ose-prometheus-node-exporter:v4.15
mgr/cephadm/container_image_alertmanager = registry.redhat.io/openshift4/ose-prometheus-alertmanager:v4.15
mgr/cephadm/container_image_grafana = registry.redhat.io/rhceph/grafana-rhel9:latest
mgr/cephadm/container_image_haproxy = registry.redhat.io/rhceph/rhceph-haproxy-rhel9:latest
mgr/cephadm/container_image_keepalived = registry.redhat.io/rhceph/keepalived-rhel9:latest
mgr/cephadm/container_image_nvmeof = registry.redhat.io/rhceph/ceph-nvmeof-rhel9:1.4
mgr/cephadm/container_image_snmp_gateway = registry.redhat.io/rhceph/snmp-notifier-rhel9:latest
mgr/cephadm/container_image_samba = cp.icr.io/cp/ibm-ceph/samba-server-rhel9:v8.1
mgr/cephadm/container_image_samba_metrics = cp.icr.io/cp/ibm-ceph/samba-metrics-rhel9:v8.1
mgr/cephadm/container_image_nginx = registry.redhat.io/rhel9/nginx-124:latest
mgr/cephadm/container_image_oauth2_proxy = registry.redhat.io/rhceph/oauth2-proxy-rhel9:v7.6.0

```
that lists out the image in a format that can be put into a ceph conf and passed to bootstrap or run through `ceph config  assimilate-conf`. Should we just have this new command functionality instead be a format flag or something on the existing command (for the binary only, we don't have anything for the mgr module yet)?

rkachach commented:
> @rkachach we have this command already in the binary
> 
> ```
> [root@vm-00 ~]# cephadm list-images
> [mgr]
> mgr/cephadm/container_image_prometheus = registry.redhat.io/openshift4/ose-prometheus:v4.15
> mgr/cephadm/container_image_loki = registry.redhat.io/openshift-logging/logging-loki-rhel8:v3.1.0
> mgr/cephadm/container_image_promtail = registry.redhat.io/rhceph/rhceph-promtail-rhel9:v3.0.0
> mgr/cephadm/container_image_node_exporter = registry.redhat.io/openshift4/ose-prometheus-node-exporter:v4.15
> mgr/cephadm/container_image_alertmanager = registry.redhat.io/openshift4/ose-prometheus-alertmanager:v4.15
> mgr/cephadm/container_image_grafana = registry.redhat.io/rhceph/grafana-rhel9:latest
> mgr/cephadm/container_image_haproxy = registry.redhat.io/rhceph/rhceph-haproxy-rhel9:latest
> mgr/cephadm/container_image_keepalived = registry.redhat.io/rhceph/keepalived-rhel9:latest
> mgr/cephadm/container_image_nvmeof = registry.redhat.io/rhceph/ceph-nvmeof-rhel9:1.4
> mgr/cephadm/container_image_snmp_gateway = registry.redhat.io/rhceph/snmp-notifier-rhel9:latest
> mgr/cephadm/container_image_samba = cp.icr.io/cp/ibm-ceph/samba-server-rhel9:v8.1
> mgr/cephadm/container_image_samba_metrics = cp.icr.io/cp/ibm-ceph/samba-metrics-rhel9:v8.1
> mgr/cephadm/container_image_nginx = registry.redhat.io/rhel9/nginx-124:latest
> mgr/cephadm/container_image_oauth2_proxy = registry.redhat.io/rhceph/oauth2-proxy-rhel9:v7.6.0
> ```
> 
> that lists out the image in a format that can be put into a ceph conf and passed to bootstrap or run through `ceph config assimilate-conf`. Should we just have this new command functionality instead be a format flag or something on the existing command (for the binary only, we don't have anything for the mgr module yet)?

Thanks @adk3798 I wasn't aware of this command existence. I added a new argument (format) so when `json` or `yaml` to get the output I'm looking for. Overall I think the default behaviour of the command is kind of "misleading" because it's very specific and list the images for mgr section but I didn't want to change this behaviour to do not break legacy software that depends on its output.

I also added the `execute_early` so command can be executed by normal users.

phlogistonjohn commented:
FYI - the code has linting and black formatting errors that is causing make check to fail.
